src/reducers/fileTree.tsx

```diff
@@ -0,0 +1,179 @@
+    }
+
+    return 1;
+  });
+
+  node.children.forEach((child) => {
+    if ('children' in child) {
+      recursiveSortInPlace(child);
+    }
+  });
+};
+
+const getVersionName = (version: Version) => {
+  return getLocalizedString(version.addon.name);
+};
+
+export const getRootPath = (version: Version) => {
+  return `root-${getVersionName(version)}`;
+};
+
+export const buildFileTree = (version: Version): DirectoryNode => {
+  const { entries } = version;
+  const root: DirectoryNode = {
+    id: getRootPath(version),
+    name: getVersionName(version),
+    children: [],
+  };
+
+  // We need to know how depth the tree is because we'll build the Treebeard
+  // tree depth by depth.
+  const maxDepth = entries.reduce((max, entry) => {
+    if (entry.depth > max) {
+      return entry.depth;
+    }
+
+    return max;
+  }, 0);
+
+  let currentDepth = 0;
+  while (currentDepth <= maxDepth) {
+    // We find the entries for the current depth.
+    const currentEntries = entries.filter(
+      // eslint-disable-next-line no-loop-func
+      (entry) => entry.depth === currentDepth,
+    );
+
+    // This is where we create new "nodes" for each entry.
+    // eslint-disable-next-line no-loop-func
+    currentEntries.forEach((entry) => {
+      let currentNode = root;
+
+      if (currentDepth > 0) {
+        // We need to find the current node (directory) to add the current
+        // entry in its children. We do this by splitting the `path` attribute
+        // and visit each node until we reach the desired node.
+        //
+        // This only applies when the current depth is not 0 (a.k.a. the root
+        // directory) because we already know `root`.
+        const parts = entry.path.split('/');
+        // Remove the filename
+        parts.pop();
+
+        for (let i = 0; i < parts.length; i++) {
+          const foundNode = currentNode.children.find(
+            (child: TreeNode) => child.name === parts[i],
+          ) as DirectoryNode;
+
+          if (foundNode) {
+            currentNode = foundNode;
+          }
+
+          // TODO: this should not happen but what if we don't find a node?
+        }
+      }
+
+      // Create a new node.
+      let node: TreeNode = {
+        id: entry.path,
+        name: entry.filename,
+      };
+
+      // When the entry is a directory, we create a `DirectoryNode`.
+      if (entry.type === 'directory') {
+        node = {
+          ...node,
+          children: [],
+        };
+      }
+
+      currentNode.children.push(node);
+    });
+
+    // To the next depth.
+    currentDepth++;
+  }
+
+  recursiveSortInPlace(root);
+
+  return root;
+};
+
+export type FileTreeState = {
+  forVersionId: void | number;
+  tree: DirectoryNode | void;
+};
+
+export const initialState: FileTreeState = {
+  forVersionId: undefined,
+  tree: undefined,
+};
+
+export const actions = {
+  buildTree: createAction('BUILD_TREE', (resolve) => {
+    return (payload: { version: Version }) => resolve(payload);
```

Comment:
I think it's better to keep actions light whenever possible since they live on the stack of all user activity. What about accepting `versionId` and `addonName` (or `rootName`)?

Reply:
I'm not sure I follow. `buildTree` needs a version to work with to build the tree. Are you suggesting we get that from the `versions` reducer, rather than passing it in? Is it okay to call a selector from `versions` when inside the `fileTree` reducer?

Reply:
Where does it need a `version`? I only saw two places:

- to build the root name: https://github.com/mozilla/addons-code-manager/blob/69cdeeebdc789e8cbc922c6a8c1ae9ba0b93feec/src/reducers/fileTree.tsx#L45
- to use `version.id` as a key https://github.com/mozilla/addons-code-manager/blob/69cdeeebdc789e8cbc922c6a8c1ae9ba0b93feec/src/reducers/fileTree.tsx#L170

Thus, I think the action payload can just be `dispatch(buildTree({ versionId: 123, rootName: 'some-name' }))`.

Am I missing some other use of `version`?

Reply:
Aren't we supposed to use the `entries` too? https://github.com/mozilla/addons-code-manager/blob/69cdeeebdc789e8cbc922c6a8c1ae9ba0b93feec/src/reducers/fileTree.tsx#L53

Reply:
Yes, the entries are kind of the point of the whole thing. 😉